autoPyTorch/pipeline/components/setup/lr_scheduler/CosineAnnealingWarmRestarts.py

```diff
@@ -52,10 +51,14 @@ def fit(self, X: Dict[str, Any], y: Any = None) -> BaseLRComponent:
         # Make sure there is an optimizer
         self.check_requirements(X, y)
 
+        # initialise required attributes for the scheduler
+        T_mult: int = 1
+        T_0: int = min(math.floor(X['epochs'] / self.n_restarts), 1)
```

Comment:
Yes it was a typo :P. I have used T_0, T_mult = (20, 1) as to what @ArlindKadra suggested. The problem this solves that in our case, we are using Hyperband and with that sometimes we have no. of epochs even less than the set value for T_0 and hence for some fidelity(Value of epochs) there never is a restart. With the above solution, we can guarantee that restarts will happen and the number of restarts is set to the value of the hyperparameter. Ofcourse, this is my idea as to how this is helping. But I am open to discuss for any improvements or changes needed.

Reply:
Ok, I got what you mean. Probably, we should discuss it more closely.
I will just put my opinion and open questions for later discussion.

1. The combination of Hyperband and warm restart seems not good (because cyclic scheduling weakens the correlation of performance in terms of various time budgets.) 
2. When you think about the objective function as `f(λ, B)` where `λ` is a configuration and `B` is a budget, preferably `f(λ, B) ≒ f(λ, B + 1)`. However, in this case, the length of 1 cycle depends on `B` (, which is not a hyperparameter), we cannot measure the accurate performance.

Reply:
@nabenabe0928 

- Right on the first point, which is why we thought of switching to the number of restarts. So, even if we could find a good `T_0` and `T_mul` it will not necessarily be good or transferable to the other fidelities because quite probably, the larger fidelity will not finish at a local minimum with those values. The number of restarts however is different, saying that 2 restarts work for a lower fidelity, makes it more transferable to a higher one and it will converge to a nicer region.
- I proposed a better solution based on the budget given to calculate the initial `T_0` given a fixed `T_mult` of 2.0, in case you and @ravinkohli also like it.
- Disagree that the original code was more accurate in multi-fidelity optimization. I would argue that the original code was worse, as you want the good hyperparameter configurations on lower fidelities, to be the same good ones on higher ones. Just the different `T_mul `and` T_0` diverge the configurations between the fidelities, which does not help with the correlation.

Reply:
@ArlindKadra 
For the third point, I did not say the original code is better, but I said the original code is more accurate in terms of the fidelity measure. Your statement is not mathematically clear, so please elaborate your opinion a bit more. 
I have to mention that Budget should not be a control parameter of hyperparameters.

Reply:
About the third point, I did not say the original code is better, but I said the original code is more accurate in terms of the fidelity measure.

I do not know how familiar you are with the successive halving theory, but we need to maintain `f(x, B) ≒ f(x, B)`.
However, in this case, the following two cases will show the quite results:
(1). Train a model with config `x` for 50 epochs (out of 100 epochs, it means we cut off at the 50 epoch).
(2). Train a model with config `x` for 50 epochs (out of 50 epochs, it means we complete a full training).
I mentioned this really matters.

Reply:
With f(x, B) ≒ f(x, B) you probably mean f(x, B) ≒ f(x, B + 1).
If so, in your opinion, what would be a closer approximate performance between 2 different budgets/fidelities, a certain T_mul and T_0 or the number of restarts?

Reply:
I agree with you that there might be a loss of correlation. I think as this is being merged into the regularisation cocktails, when we merge this to the refactor_development branch we can have a discussion about it